<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" >
    <xacro:include filename="inertial_macros.xacro" />
    <material name="white">
        <color rgba="1 1 1 1" />
    </material>

    <material name="orange">
        <color rgba="1 0.3 0.1 1"/>
    </material>

    <material name="blue">
        <color rgba="0.2 0.2 1 1"/>
    </material>

    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>

    <material name="red">
        <color rgba="1 0 0 1"/>
    </material>

    <!-- BASE LINK -->

    <link name="base_link">

    </link>

    <joint name="chassis_joint" type="fixed">
        <parent link="base_link"/>
        <child link="chassis"/>
        <origin xyz="0 0 0"/>
    </joint>

    <link name="chassis">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/body.dae"/>
            </geometry>
            <material name="white"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/body.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="5.0" x="0.75" y="0.75" z="0.25">
            <origin xyz="0 0 0.125" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>

    <gazebo reference="chassis">
        <material>Gazebo/Blue</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- LEFT WHEEL -->
    <!-- Front -->
    <joint name="left_front_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="left_front_wheel"/>
        <origin xyz="0.3 0.3975 0" rpy="-${pi/2} 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <link name="left_front_wheel">
        <visual>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.045" radius="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="left_front_wheel">
        <material>Gazebo/Black</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>
    
    <!-- Back -->
    <joint name="left_back_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="left_back_wheel"/>
        <origin xyz="-0.3 0.3975 0" rpy="-${pi/2} 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <link name="left_back_wheel">
        <visual>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.045" radius="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="left_back_wheel">
        <material>Gazebo/Black</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- RIGHT WHEEl -->
    <!-- Front -->
    <joint name="right_front_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="right_front_wheel"/>
        <origin xyz="0.3 -0.3975 0" rpy="${pi/2} 0 0"/>
        <axis xyz="0 0 -1"/>
    </joint>

    <link name="right_front_wheel">
        <visual>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.045" radius="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="right_front_wheel">
        <material>Gazebo/Black</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>
    
    <!-- Back -->
    <joint name="right_back_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="right_back_wheel"/>
        <origin xyz="-0.3 -0.3975 0" rpy="${pi/2} 0 0"/>
        <axis xyz="0 0 -1"/>
    </joint>

    <link name="right_back_wheel">
        <visual>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="0.045" radius="0.05" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.045" radius="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="right_back_wheel">
        <material>Gazebo/Black</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>
    
    <!-- ROBOT ARM-->
    <!-- Arm Base-->
    <joint name="base_joint" type="fixed">
        <parent link="chassis"/>
        <child link="base"/>
        <origin xyz="0 0 0.25"/>
    </joint>
    
    <link name="base">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_base.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_base.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="1.0" length="0.06" radius="0.12">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="base">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- Arm Joint 1 -->
    <joint name="arm_1_joint" type="revolute">
        <parent link="base"/>
        <child link="arm_1"/>
        <origin xyz="0 0 0.06" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower = "-3.14" upper = "3.14" effort = "1" velocity="1" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="arm_1">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_1.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_1.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="2.0" length="0.16" radius="0.05">
            <origin xyz="0 0 0.075" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="arm_1">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>
    
    <!-- Arm Joint 2 -->

    <joint name="arm_2_joint" type="revolute">
        <parent link="arm_1"/>
        <child link="arm_2"/>
        <origin xyz="0.071 0 0.105" rpy="0 0 -${pi/2}"/>
        <axis xyz="0 1 0"/>
        <limit lower = "-1.74" upper = "1.74" effort = "1000000000000" velocity="5" />\
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="arm_2">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_2.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_2.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.565" radius="0.05">
            <origin xyz="0 0.065 0.2325" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="arm_2">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- Arm Joint 3 -->

    <joint name="arm_3_joint" type="revolute">
        <parent link="arm_2"/>
        <child link="arm_3"/>
        <origin xyz="0 0 0.455" rpy="0 0 -${pi/2}"/>
        <axis xyz="1 0 0"/>
        <limit lower = "-2.88" upper = "2.88" effort = "10" velocity="1" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="arm_3">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_3.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_3.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.537" radius="0.04">
            <origin xyz="0.06 0 0.2125" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="arm_3">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- Arm Joint 4 -->

    <joint name="arm_4_joint" type="revolute">
        <parent link="arm_3"/>
        <child link="arm_4"/>
        <origin xyz="0 0 0.437" rpy="-${pi} 0 -${pi}"/>
        <axis xyz="1 0 0"/>
        <limit lower = "-3.14" upper = "3.14" effort = "1" velocity="1" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="arm_4">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_4.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_4.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.115" radius="0.045">
            <origin xyz="0.06 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="arm_4">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- Arm Joint 5 -->

    <joint name="arm_5_joint" type="revolute">
        <parent link="arm_4"/>
        <child link="arm_5"/>
        <origin xyz="0.06 0 -0.06" rpy="-${pi/2} 0 0"/>
        <axis xyz="0 1 0"/>
        <limit lower = "-3.14" upper = "3.14" effort = "1" velocity="1" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="arm_5">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_5.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Arm_joint_5.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.105" radius="0.045">
            <origin xyz="0 0.06 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="arm_5">
        <material>Gazebo/Orange</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- GRIP -->
    <!-- Grip 1 -->

    <joint name="grip_1_joint" type="revolute">
        <parent link="arm_5"/>
        <child link="grip_1"/>
        <origin xyz="-0.06 0.06 0" rpy="0 0 0"/>
        <axis xyz="1 0 0"/>
        <limit lower = "-1.74" upper = "1.74" effort = "1" velocity="0.01" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="grip_1">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Grip_1.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Grip_1.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.025" radius="0.035">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="grip_1">
        <material>Gazebo/Red</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- Grip 2 -->

    <joint name="grip_2_joint" type="revolute">
        <parent link="arm_5"/>
        <child link="grip_2"/>
        <origin xyz="-0.075 0.06 0" rpy="${pi} 0 -${pi}"/>
        <axis xyz="-1 0 0"/>
        <limit lower = "-1.74" upper = "1.74" effort = "1" velocity="0.01" />
        <dynamics damping="1" friction="1"/>
    </joint>

    <link name="grip_2">
        <visual>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Grip_2.dae"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file:///home/white/final_ws/src/articubot_one/description/Grip_2.dae"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.025" radius="0.035">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="grip_2">
        <material>Gazebo/Red</material>
        <mu1 value="0.1"/>
        <mu2 value="0.1"/>
    </gazebo>

    <!-- LIDAR -->

    <joint name="laser_joint" type="fixed">
        <parent link="chassis"/>
        <child link="laser_frame"/>
        <origin xyz="0.25 0 0.25" rpy="0 0 0"/>
    </joint>
 
    <link name="laser_frame">
        <visual>
            <geometry>
                <cylinder radius="0.05" length="0.005"/>
            </geometry>
            <material name="Black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.05" length="0.005"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.005" radius="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>
 
 
 
    <gazebo reference="laser_frame">
        <material>Gazebo/Black</material>
 
        <sensor name="laser" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>10</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>360</samples>
                        <min_angle>-3.14</min_angle>
                        <max_angle>3.14</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.5</min>
                    <max>12</max>
                </range>
            </ray>
            
            <plugin name="laser_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=scan</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>laser_frame</frame_name>
            </plugin>
        </sensor>
    </gazebo>

</robot>